app/models/spree/credit_card_decorator.rb

```diff
@@ -25,4 +28,16 @@ def has_payment_profile? # rubocop:disable Style/PredicateName
   def save_requested_by_customer?
     !!@save_requested_by_customer
   end
+
+  private
+
+  def default_missing?
+    user.credit_cards.where(is_default: true).none?
+  end
+
+  def ensure_default
+    return unless user
+    return unless is_default? || default_missing?
+    user.credit_cards.update_all(['is_default=(id=?)', id])
+  end
```

Comment:
Unless we don't add the constraint at DB level nothing stops us from having two cards with `is_default = true` due to race conditions. Adding a unique index on `is_default`for true values will do it. Check https://stackoverflow.com/a/28167225

Reply:
Trying to think of a better way of achieving this. If I add a unique constraint on `user_id` and `is_default` then I can only have a maximum of two cards per user, because a third card (ie. the second with a `is_default=false` for the user) will break the constraint, right? Or does `false` behave like `null` in unique indexes, in that they don't break the constraint? I will check...

Reply:
Looks like it is not possible to create partial indexes until rails `v4.0`...there is no support for the `where` option in `v3.2.22`

I'm trying to understand the mechanism by which we would end up with multiple card using this logic. There is only one query. Are you suggesting that if two queries are submitted to postgres at the same time by two separate workers, that they may write to the same set of rows at the same time, leaving them in a mess? I don't understand the full details of postgres' row locking, but I would have thought that this was protected against. If not, I think the best solution is probably a table lock? That way we can be sure that only one query is being executed by postgres at a time.

Reply:
I think it's better to avoid any issues like that with a better database design. So one user can have one default card. The standard way is to store the default card id as a column on the user table. But if we did that with everything, the user table become huge and we don't always need to know the user's default card when we load a user. So the alternative is to create a new table with the columns `user_id` and `card_id`. In this case you can make both of them unique via an index, because a user should have only one default and a card belongs to only one user. This approach also avoids changing a Spree table. It also means that the lookup of the default card of a user is completely indexed. Otherwise the database has to look at all cards of a user and filter by the `is_default` column.

Reply:
I like that approach, but I'm not sure that Rails is smart enough to magically manage a join table as a `has_one` association in both directions. We might need to add a model for the join table so that we can just use `has_one, through`. `UserDefaultCard`?

Reply:
Hm, I'm playing around with this idea and it seems to be resulting in much more pollution of the user model. I'll persist and see if I can pull it back a bit.

Reply:
@sauloperez @mkllnk

Ok, I've had a really good crack at implementing Maikel's idea, but I am really fighting rails pretty hard. Doing a join table for a `has_one` just doesn't seem to be working for me. This is where I've ended up: override the setter method on a `has_one through` association.😱 I got here because the default implementation of `default_card=`doesn't update the record in the join table, it just creates a new one and nulls the `user_id` on the old record. 

````ruby
class UsersDefaultCard < ActiveRecord::Base
  belongs_to :user, class_name: 'Spree::User'
  belongs_to :default_card, class_name: 'Spree::CreditCard'
end

Spree.user_class.class_eval do
 ...
  has_one :users_default_card
  has_one :default_card, through: :users_default_card
 ...
  def default_card=(card)
    default = if users_default_card
      users_default_card
    else
      build_users_default_card
    end
    default.update_attributes(default_card_id: card.id)
    self[:users_default_card] = default
    self[:default_card] = card
  end
...
end
````

This doesn't feel like an improvement to me, any suggestions?

Reply:
I wouldn't go that way then. Let's start with what you implemented.

> I'm trying to understand the mechanism by which we would end up with multiple default cards using this logic.

If the DB doesn't enforce the data integrity chances are that the data will end up in a bad state due to [race conditions](https://en.wikipedia.org/wiki/Race_condition).

Reply:
I know what a race condition is, I am just trying to understand how we would actually end up with one, so that I can try to prevent it further up the chain. IMO db constraints should really be a last resort protection against race conditions, not the primary way they are dealt with.

Reply:
I cannot think of any other way other than DB constraints. You can't control how and when the requests are processed by the unicorn workers.

Reply:
@mkllnk I had a go at your suggestion but I really couldn't get it to work easily. Mainly I couldn't work out how to set a has_one association using `update_attributes`.

Are you both ok if we go with the existing version, given that there doesn't seem to be an obvious way of adding an appropriate db-level constraint?

Reply:
Yes, that's okay. As Kirsten said the other day: perfect is the enemy of done. It's very unlikely that this causes a problem. I mean it's unlikely that somebody tries to set two different cards as default within seconds and then complains about the wrong card being charged.